share/shwap/p2p/bitswap/block_fetch.go

```diff
@@ -0,0 +1,222 @@
+
+	"github.com/celestiaorg/celestia-node/share"
+	bitswappb "github.com/celestiaorg/celestia-node/share/shwap/p2p/bitswap/pb"
+)
+
+// WithFetchSession instantiates a new Fetch session and saves it on the context.
+// Session reuses peers known to have Blocks without rediscovering.
+func WithFetchSession(ctx context.Context, exchg exchange.SessionExchange) context.Context {
+	fetcher := exchg.NewSession(ctx)
+	return context.WithValue(ctx, fetcherKey, fetcher)
+}
+
+// Fetch fetches and populates given Blocks using Fetcher wrapping Bitswap.
+//
+// Validates Block against the given Root and skips Blocks that are already populated.
+// Gracefully synchronize identical Blocks requested simultaneously.
+// Blocks until either context is canceled or all Blocks are fetched and populated.
+func Fetch(ctx context.Context, exchg exchange.Interface, root *share.Root, blks ...Block) error {
+	for from, to := 0, 0; to < len(blks); { //nolint:wastedassign // it's not actually wasted
+		from, to = to, to+maxPerFetch
+		if to >= len(blks) {
+			to = len(blks)
+		}
+
+		err := fetch(ctx, exchg, root, blks[from:to]...)
+		if err != nil {
+			return err
+		}
+	}
+
+	return ctx.Err()
+}
+
+// maxPerFetch sets the limit for maximum items in a single fetch.
+// It's a heuristic coming from Bitswap, which apparently can't process more than ~1024 in a single
+// GetBlock call. Going beyond that stalls the call indefinitely.
+const maxPerFetch = 1024
+
+// fetch fetches given Blocks.
+// See [Fetch] for detailed description.
+func fetch(ctx context.Context, exchg exchange.Interface, root *share.Root, blks ...Block) error {
+	fetcher := getFetcher(ctx, exchg)
+	cids := make([]cid.Cid, 0, len(blks))
+	duplicates := make(map[cid.Cid]Block)
+	for _, blk := range blks {
+		if !blk.IsEmpty() {
+			continue // skip populated Blocks
+		}
+
+		cid := blk.CID() // memoize CID for reuse as it ain't free
+		cids = append(cids, cid)
+
+		// store the UnmarshalFn s.t. hasher can access it
+		// and fill in the Block
+		unmarshalFn := blk.UnmarshalFn(root)
+		_, exists := unmarshalFns.LoadOrStore(cid, &unmarshalEntry{UnmarshalFn: unmarshalFn})
+		if exists {
+			// the unmarshalFn has already been stored for the cid
+			// means there is ongoing fetch happening for the same cid
+			duplicates[cid] = blk // so mark the Block as duplicate
+		} else {
+			// cleanup are by the original requester and
+			// only after we are sure we got the block
+			defer unmarshalFns.Delete(cid)
+		}
+	}
+
+	blkCh, err := fetcher.GetBlocks(ctx, cids)
+	if err != nil {
+		return fmt.Errorf("requesting Bitswap blocks: %w", err)
+	}
+
+	for bitswapBlk := range blkCh { // GetBlocks closes blkCh on ctx cancellation
+		if err := exchg.NotifyNewBlocks(ctx, bitswapBlk); err != nil {
+			log.Error("failed to notify new Bitswap blocks: %w", err)
+		}
+
+		blk, ok := duplicates[bitswapBlk.Cid()]
+		if !ok {
+			// common case: the block was populated by the hasher, so skip
+			continue
+		}
```

Comment:
I was thinking about the loop that checks if all the blocks arent empty. It will give the user assurance that all the blocks are populated leaving the Fetch call. Lets add it

Reply:
Added